File: CS/Dennis.Search.Win/SearchObjectBase.cs
using DevExpress.ExpressApp.Model;
using System;
using DevExpress.Xpo;
using System.Collections;
using System.ComponentModel;
using DevExpress.Data.Filtering;
using DevExpress.ExpressApp.Model.NodeWrappers;

namespace Dennis.Search.Win {
    public interface ISearchObject {
        event EventHandler<SearchStartEventArgs> SearchStart;
        event EventHandler<SearchCompleteEventArgs> SearchComplete;
        GroupOperator Criteria { get;set;}
        ICollection SearchResults { get;}
        bool IsSearching { get;}
        void Search();
        void Reset();
    }
    public class SearchStartEventArgs : CancelEventArgs {
        private CriteriaOperator criteriaCore;
        public SearchStartEventArgs(CriteriaOperator criteria) {
            criteriaCore = criteria;
        }
        public CriteriaOperator Criteria {
            get { return criteriaCore; }
            set { criteriaCore = value; }
        }
    }
    public class SearchCompleteEventArgs : EventArgs {
        private ICollection searchResultsCore;
        private Exception errorCore;
        public SearchCompleteEventArgs(ICollection searchResults, Exception error) {
            searchResultsCore = searchResults;
            errorCore = error;
        }
        public Exception Error {
            get { return errorCore; }
        }
        public ICollection SearchResults {
            get { return searchResultsCore; }
        }
    }
    [NonPersistent]
    public abstract class SearchObjectBase<T> : XPBaseObject, ISearchObject where T : IXPSimpleObject {
        private XPCollection<T> searchResultsCore;
        private GroupOperator criteriaCore;
        private bool isSearchingCore;
        protected const string SearchObjectPropertyEditorFullName = "Dennis.Search.Win.Editors";
        protected const string SearchResultsCollectionName = "SearchResults";
        protected const BinaryOperatorType BinaryCriteriaType = BinaryOperatorType.Like;
        protected const GroupOperatorType GroupCriteriaType = GroupOperatorType.And;
        protected readonly CriteriaOperator EmptyCollectionCriteria = CriteriaOperator.Parse("1=0");
        public SearchObjectBase(Session session) : base(session) { }
        [ModelDefault("PropertyEditorType", SearchObjectPropertyEditorFullName)]
        public GroupOperator Criteria {
            get {
                if (ReferenceEquals(criteriaCore, null))
                    criteriaCore = new GroupOperator(GroupCriteriaType, EmptyCollectionCriteria);
                return criteriaCore;
            }
            set { SetPropertyValue("Criteria", ref criteriaCore, value); }
        }
        protected override void OnChanged(string propertyName, object oldValue, object newValue) {
            base.OnChanged(propertyName, oldValue, newValue);
            if (!IsLoading && !IsSaving)
                if (propertyName != "Criteria" && propertyName != SearchResultsCollectionName)
                    UpdateCriteria(propertyName, newValue);
        }
        protected virtual void UpdateCriteria(string propertyName, object propertyValue) {
            bool updated = false;
            foreach (CriteriaOperator operand in Criteria.Operands) {
                BinaryOperator binaryOperator = operand as BinaryOperator;
                if (!ReferenceEquals(binaryOperator, null) && !ReferenceEquals(binaryOperator, EmptyCollectionCriteria))
                    if (((OperandProperty)binaryOperator.LeftOperand).PropertyName == propertyName) {
                        ((OperandValue)binaryOperator.RightOperand).Value = propertyValue;
                        updated = true;
                        break;
                    }
            }
            if (!updated)
                if (propertyValue is String) {
                    Criteria.Operands.Add(new BinaryOperator(propertyName, propertyValue, BinaryCriteriaType));
                } else {
                    Criteria.Operands.Add(new BinaryOperator(propertyName, propertyValue));
                }
        }
        [Browsable(false)]
        public bool IsSearching {
            get { return isSearchingCore; }
        }
        public XPCollection<T> SearchResults {
            get {
                if (searchResultsCore == null)
                    searchResultsCore = new XPCollection<T>(Session, EmptyCollectionCriteria, null);
                return searchResultsCore;
            }
        }
        public virtual void Reset() {
            criteriaCore = null;
            searchResultsCore.Criteria = EmptyCollectionCriteria;
        }
        ICollection ISearchObject.SearchResults {
            get { return SearchResults; }
        }
        public virtual void Search() {
            isSearchingCore = true;
            if (Criteria.Operands.Contains(EmptyCollectionCriteria))
                Criteria.Operands.Remove(EmptyCollectionCriteria);
            SearchStartEventArgs args = new SearchStartEventArgs(Criteria);
            OnSearchStart(args);
            if (!args.Cancel) {
                searchResultsCore.SuspendChangedEvents();
                searchResultsCore.Criteria = args.Criteria;
                searchResultsCore.LoadAsync(LoadSearchResultsCallback);
            }
        }
        private void LoadSearchResultsCallback(ICollection[] result, Exception exception) {
            try {
                if (exception != null)
                    throw new InvalidOperationException(string.Format("Cannot load search results for the {0} object type due to an error: {1}", typeof(T), exception.Message));
            } finally {
                isSearchingCore = false;
                OnSearchComplete(new SearchCompleteEventArgs(result != null && result.Length > 0 ? result[0] : null, exception));
                searchResultsCore.ResumeChangedEvents();
            }
        }
        protected virtual void OnSearchStart(SearchStartEventArgs args) {
            if (SearchStart != null)
                SearchStart(this, args);
        }
        protected virtual void OnSearchComplete(SearchCompleteEventArgs args) {
            if (SearchComplete != null)
                SearchComplete(this, args);
        }
        public event EventHandler<SearchStartEventArgs> SearchStart;
        public event EventHandler<SearchCompleteEventArgs> SearchComplete;
    }
}

File: VB/Dennis.Search.Win/SearchObjectBase.vb
Imports DevExpress.ExpressApp.Model
Imports System
Imports DevExpress.Xpo
Imports System.Collections
Imports System.ComponentModel
Imports DevExpress.Data.Filtering
Imports DevExpress.ExpressApp.Model.NodeWrappers

Namespace Dennis.Search.Win
    Public Interface ISearchObject
        Event SearchStart As EventHandler(Of SearchStartEventArgs)
        Event SearchComplete As EventHandler(Of SearchCompleteEventArgs)
        Property Criteria() As GroupOperator
        ReadOnly Property SearchResults() As ICollection
        ReadOnly Property IsSearching() As Boolean
        Sub Search()
        Sub Reset()
    End Interface
    Public Class SearchStartEventArgs
        Inherits CancelEventArgs

        Private criteriaCore As CriteriaOperator
        Public Sub New(ByVal criteria As CriteriaOperator)
            criteriaCore = criteria
        End Sub
        Public Property Criteria() As CriteriaOperator
            Get
                Return criteriaCore
            End Get
            Set(ByVal value As CriteriaOperator)
                criteriaCore = value
            End Set
        End Property
    End Class
    Public Class SearchCompleteEventArgs
        Inherits EventArgs

        Private searchResultsCore As ICollection
        Private errorCore As Exception
        Public Sub New(ByVal searchResults As ICollection, ByVal [error] As Exception)
            searchResultsCore = searchResults
            errorCore = [error]
        End Sub
        Public ReadOnly Property [Error]() As Exception
            Get
                Return errorCore
            End Get
        End Property
        Public ReadOnly Property SearchResults() As ICollection
            Get
                Return searchResultsCore
            End Get
        End Property
    End Class
    <NonPersistent> _
    Public MustInherit Class SearchObjectBase(Of T As IXPSimpleObject)
        Inherits XPBaseObject
        Implements ISearchObject

        Private searchResultsCore As XPCollection(Of T)
        Private criteriaCore As GroupOperator
        Private isSearchingCore As Boolean
        Protected Const SearchObjectPropertyEditorFullName As String = "Dennis.Search.Win.Editors"
        Protected Const SearchResultsCollectionName As String = "SearchResults"
        Protected Const BinaryCriteriaType As BinaryOperatorType = BinaryOperatorType.Like
        Protected Const GroupCriteriaType As GroupOperatorType = GroupOperatorType.And
        Protected ReadOnly EmptyCollectionCriteria As CriteriaOperator = CriteriaOperator.Parse("1=0")
        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub
        <ModelDefault("PropertyEditorType", SearchObjectPropertyEditorFullName)> _
        Public Property Criteria() As GroupOperator Implements ISearchObject.Criteria
            Get
                If ReferenceEquals(criteriaCore, Nothing) Then
                    criteriaCore = New GroupOperator(GroupCriteriaType, EmptyCollectionCriteria)
                End If
                Return criteriaCore
            End Get
            Set(ByVal value As GroupOperator)
                SetPropertyValue("Criteria", criteriaCore, value)
            End Set
        End Property
        Protected Overrides Sub OnChanged(ByVal propertyName As String, ByVal oldValue As Object, ByVal newValue As Object)
            MyBase.OnChanged(propertyName, oldValue, newValue)
            If Not IsLoading AndAlso Not IsSaving Then
                If propertyName <> "Criteria" AndAlso propertyName <> SearchResultsCollectionName Then
                    UpdateCriteria(propertyName, newValue)
                End If
            End If
        End Sub
        Protected Overridable Sub UpdateCriteria(ByVal propertyName As String, ByVal propertyValue As Object)
            Dim updated As Boolean = False
            For Each operand As CriteriaOperator In Criteria.Operands
                Dim binaryOperator As BinaryOperator = TryCast(operand, BinaryOperator)
                If Not ReferenceEquals(binaryOperator, Nothing) AndAlso Not ReferenceEquals(binaryOperator, EmptyCollectionCriteria) Then
                    If CType(binaryOperator.LeftOperand, OperandProperty).PropertyName = propertyName Then
                        CType(binaryOperator.RightOperand, OperandValue).Value = propertyValue
                        updated = True
                        Exit For
                    End If
                End If
            Next operand
            If Not updated Then
                If TypeOf propertyValue Is String Then
                    Criteria.Operands.Add(New BinaryOperator(propertyName, propertyValue, BinaryCriteriaType))
                Else
                    Criteria.Operands.Add(New BinaryOperator(propertyName, propertyValue))
                End If
            End If
        End Sub
        <Browsable(False)> _
        Public ReadOnly Property IsSearching() As Boolean Implements ISearchObject.IsSearching
            Get
                Return isSearchingCore
            End Get
        End Property
        Public ReadOnly Property SearchResults() As XPCollection(Of T)
            Get
                If searchResultsCore Is Nothing Then
                    searchResultsCore = New XPCollection(Of T)(Session, EmptyCollectionCriteria, Nothing)
                End If
                Return searchResultsCore
            End Get
        End Property
        Public Overridable Sub Reset() Implements ISearchObject.Reset
            criteriaCore = Nothing
            searchResultsCore.Criteria = EmptyCollectionCriteria
        End Sub
        Private ReadOnly Property ISearchObject_SearchResults() As ICollection Implements ISearchObject.SearchResults
            Get
                Return SearchResults
            End Get
        End Property
        Public Overridable Sub Search() Implements ISearchObject.Search
            isSearchingCore = True
            If Criteria.Operands.Contains(EmptyCollectionCriteria) Then
                Criteria.Operands.Remove(EmptyCollectionCriteria)
            End If
            Dim args As New SearchStartEventArgs(Criteria)
            OnSearchStart(args)
            If Not args.Cancel Then
                searchResultsCore.SuspendChangedEvents()
                searchResultsCore.Criteria = args.Criteria
                searchResultsCore.LoadAsync(AddressOf LoadSearchResultsCallback)
            End If
        End Sub
        Private Sub LoadSearchResultsCallback(ByVal result() As ICollection, ByVal exception As Exception)
            Try
                If exception IsNot Nothing Then
                    Throw New InvalidOperationException(String.Format("Cannot load search results for the {0} object type due to an error: {1}", GetType(T), exception.Message))
                End If
            Finally
                isSearchingCore = False
                OnSearchComplete(New SearchCompleteEventArgs(If(result IsNot Nothing AndAlso result.Length > 0, result(0), Nothing), exception))
                searchResultsCore.ResumeChangedEvents()
            End Try
        End Sub
        Protected Overridable Sub OnSearchStart(ByVal args As SearchStartEventArgs)
            RaiseEvent SearchStart(Me, args)
        End Sub
        Protected Overridable Sub OnSearchComplete(ByVal args As SearchCompleteEventArgs)
            RaiseEvent SearchComplete(Me, args)
        End Sub
        Public Event SearchStart As EventHandler(Of SearchStartEventArgs) Implements ISearchObject.SearchStart
        Public Event SearchComplete As EventHandler(Of SearchCompleteEventArgs) Implements ISearchObject.SearchComplete
    End Class
End Namespace